File: 200-number-of-islands/200-number-of-islands.java
class Solution {
    static boolean[][] visited;
	static int m;
	static int n;
	static int[] dx = {1, -1, 0, 0};
	static int[] dy = {0, 0, 1, -1};
	static boolean isOutOfBound(int r, int c) {
		if(r<0 || r>=m || c<0 || c>=n) return true;
		return false;
	}
	static void dfs(int r, int c, char[][] grid) {
		if(grid[r][c]=='0') return;
		
		// 1 이면 계속 탐색
		for(int i=0; i<4; i++) {
			int nr = r + dx[i];
			int nc = c + dy[i];
			if(!isOutOfBound(nr, nc) && !visited[nr][nc] && grid[nr][nc]=='1') {
				visited[nr][nc]= true;
				dfs(nr, nc, grid);
			}
		}
	}
    public int numIslands(char[][] grid) {
        m = grid.length;
		n = grid[0].length;
		visited = new boolean[m][n];
		
		int cnt = 0;
		for(int i=0; i<m; i++) {
			for(int j=0; j<n; j++) {
				if(grid[i][j]=='1' && !visited[i][j]) {
					visited[i][j] = true;
					dfs(i, j, grid);
					cnt ++;
				}
			}
		}
		
		return cnt;
    }
}

File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dx = [-1,1, 0, 0]
        dy = [0, 0, 1, -1]
        def dfs(i, j):
            if i<0 or i >= len(grid) or j<0 or j >= len(grid[0]) or grid[i][j]!='1':
                return
            grid[i][j]=0
            # 동서남북 탐색
            for dir in range(4):
                x = i + dx[dir]
                y = j + dy[dir]
                dfs(x,y)

        cnt = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] =='1':
                    dfs(i, j)
                    # 모든 육지 탐색 후 카운트 1증가
                    cnt += 1
        return cnt